superset-frontend/src/views/CRUD/hooks.ts

```diff
@@ -299,64 +299,66 @@ export function useSingleViewResource<D extends object = any>(
   };
 }
 
-// the hooks api has some known limitations around stale state in closures.
-// See https://github.com/reactjs/rfcs/blob/master/text/0068-react-hooks.md#drawbacks
-// the useRef hook is a way of getting around these limitations by having a consistent ref
-// that points to the most recent value.
 export function useFavoriteStatus(
-  initialState: FavoriteStatus,
-  baseURL: string,
+  type: 'chart' | 'dashboard',
+  ids: Array<string | number>,
   handleErrorMsg: (message: string) => void,
 ) {
-  const [favoriteStatus, setFavoriteStatus] = useState<FavoriteStatus>(
-    initialState,
-  );
-  const favoriteStatusRef = useRef<FavoriteStatus>(favoriteStatus);
-  useEffect(() => {
-    favoriteStatusRef.current = favoriteStatus;
-  });
+  const [favoriteStatus, setFavoriteStatus] = useState<FavoriteStatus>({});
 
   const updateFavoriteStatus = (update: FavoriteStatus) =>
     setFavoriteStatus(currentState => ({ ...currentState, ...update }));
 
-  const fetchFaveStar = (id: number) => {
+  useEffect(() => {
+    if (!ids.length) {
+      return;
+    }
     SupersetClient.get({
-      endpoint: `${baseURL}/${id}/count/`,
+      endpoint: `/api/v1/${type}/favorite_status/?q=${rison.encode(ids)}`,
     }).then(
       ({ json }) => {
-        updateFavoriteStatus({ [id]: json.count > 0 });
+        const update = (json?.result as {
+          id: string;
+          value: boolean;
+        }[])?.reduce((acc, element) => {
+          acc[element.id] = element.value;
+          return acc;
+        }, {});
+        updateFavoriteStatus(update);
       },
       createErrorHandler(errMsg =>
         handleErrorMsg(
           t('There was an error fetching the favorite status: %s', errMsg),
         ),
       ),
     );
-  };
-
-  const saveFaveStar = (id: number, isStarred: boolean) => {
-    const urlSuffix = isStarred ? 'unselect' : 'select';
-
-    SupersetClient.get({
-      endpoint: `${baseURL}/${id}/${urlSuffix}/`,
-    }).then(
-      () => {
-        updateFavoriteStatus({ [id]: !isStarred });
-      },
-      createErrorHandler(errMsg =>
-        handleErrorMsg(
-          t('There was an error saving the favorite status: %s', errMsg),
+  }, [ids]);
+
+  const saveFaveStar = useCallback(
+    (id: number, isStarred: boolean) => {
+      const urlSuffix = isStarred ? 'unselect' : 'select';
+      console.log('fetching', id, isStarred);
+      SupersetClient.get({
+        endpoint: `/superset/favstar/${
+          type === 'chart' ? 'slice' : 'Dashboard'
+        }/${id}/${urlSuffix}/`,
+      }).then(
+        ({ json }) => {
+          updateFavoriteStatus({
+            [id]: (json as { count: number })?.count > 0,
+          });
+        },
+        createErrorHandler(errMsg =>
+          handleErrorMsg(
+            t('There was an error saving the favorite status: %s', errMsg),
+          ),
```

Comment:
Not a blocker but may I suggest utilizing `makeApi` from `@superset-ui/core` and add strong typing for your API?

https://github.com/apache-superset/superset-ui/blob/46728e8f0c40961f80ba2f093a1cf2248cd49eb1/packages/superset-ui-core/src/query/api/v1/index.ts#L19-L33

Reply:
I’m not aware of this util, thanks for the tip. I’ve been meaning to ask about adding type params for SupersetClient responses

Reply:
I really like the `makeApi` util!